text/0000-serialization.md

```diff
@@ -0,0 +1,394 @@
+- SDK PR: [#0000](https://github.com/Unity-Technologies/com.unity.multiplayer.mlapi/pull/0000)
+
+# Summary
+
+[summary]: #summary
+
+This feature describes our approach to serialization. The goals of the serialization feature are that it is:
+
+- Fast
+- Allocation-free
+- Consistent
+- Debuggable
+- Transparent (that is, easy for end users to follow and understand)
+- Maintainable
+- Future-ready (i.e., for jobs)
+
+This feature also proposes a new way of defining messages that supports these goals.
+
+# Motivation
+
+[motivation]: #motivation
+
+This is both a performance and architectural change to MLAPI. Currently, serialization is one of our biggest offenders for CPU usage, and we've created pools to avoid allocations. Additionally, sending a message passes through a number of steps that can be difficult to follow for an unfamiliar user - the message is first serialized (and serialized differently depending on whether it's going to another node or going to the inbound queue), placed into the MessageQueueHistoryFrame as a buffer, then later deserialized, modified, and reserialized in a different format. The serialization itself is also sometimes spread out throughout the code base - one function serializes some of it, then calls another function that adds more to it, and so on. All of this comes at the cost that users unfamiliar with the code base can have a difficult time following the path a message takes or understanding what is (or should be) in the buffer in order to debug it.
+
+This feature will rethink the whole flow of serialization, from the beginning (defining messages) all the way through serializing, batching, sending, deserializing, and handling messages.
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+## Defining Messages
+
+Messages in MLAPI are represented in-memory using structs containing, as much as possible, only unmanaged POD. Each struct must implement the IMessage interface, which is defined as follows:
+
+```C#
+interface IMessage
+{
+    enum MessageType
+    {
+        //...
+    }
+
+    // Returns the type of the message, used in automatic message header generation
+    MessageType GetMessageType();
+
+    // Returns an upper bound on the message size.
+    // Doesn't have to be exactly accurate to the serialized size
+    // (over-estimating is preferred to using more CPU cycles for an exact result)
+    // but must be AT LEAST the message size and AT MOST (MTU - 6) bytes.
+    int GetSizeUpperBound();
+
+    // Serialize to a buffer
+    void Serialize(ref FastNetworkWriter writer, in NetworkContext context);
+    
+    // Handle the message
+    void Handle(in NetworkContext context);
+}
+```
+
+Additionally, each message should have a method on it as follows:
+
+```
+[MessageHandler(IMessage.MessageType.MyMessageType)]
+static void Receive(ref FastNetworkReader reader, in NetworkContext context);
```

Comment:
I do not understand why we need ILPP here for something like this. For Server/Client RPC we needed this because we wanted to rewrite how c# code behaves before we pass it to the compiler. Looks like this is just creates a lookup table from MessageType to Action<ref FastNetworkReader, in NetworkContext>.

We can just construct this lookup at runtime initialization with a Dictionary<MessageType, Action<...>>. This will result in roughly the same performance but reduces the amount of ILPP in our codebase.

Reply:
We may not need ILPP. My reason for proposing it is that I was concerned about iOS, since they don't allow the use of JIT. My concern is that if we never call these functions directly and we only discover them at runtime via reflection, that the AOT compiler would determine they're unused and not emit code for them, and then things would crash on iOS when we tried to call those functions and the JIT compiler wasn't available to compile them.

This may be an entirely invalid concern, in which case, yes, it would be much simpler to discover them at runtime.